### PR TITLE
Definition namespaces and providers

### PR DESCRIPTION
**Pull request by ml-evs**
This PR begins the work on #328, adding prefixes that can be shared across providers. I will try to follow this rough plan with my work:

- [x] Define "definition provider" [working title] 
- [x] Replace relevant mentions of "database-provider-specific" with "namespace-specific"
- [ ] ~Potentially extend the provider/link definition to add a specialism for this kind of provider~ (decided against it)
- [x] Describe mechanism for becoming a definition provider
- [x] Provide example cases

Then outside of the specification we need to:
- [x] Set up an example definition provider, as a separate repository, that can be modified for future domain-specific extensions

### CODE REVIEW

File: .words.lst [hunk 1]
```diff
@@ -113,6 +113,7 @@ enum
 exclusiveMaximum
 exclusiveMinimum
 exmpl
+fjeiwoj
```
Comment (ml-evs) [inline .words.lst:116]: This was required for me as `cspell` flagged the random invalid field name example up, whereas aspell is fine with it.

**Comment by ml-evs** [comment]
I think this is ready for some initial feedback. I've also made the start of a namespace repo at https://github.com/ml-evs/optimade-stability-namespace (currently forked to https://github.com/Materials-Consortia/optimade-stability-namespace) where I am trying to use @rartino's processing tools from #445 to generate the website/schemas (currently with no luck).

File: optimade.rst [hunk 2]
```diff
@@ -367,26 +379,61 @@ A few suggestions and mandatory requirements of the OPTIMADE specification are s
...
+Database-provider-specific fields SHOULD be fully described at the relevant :endpoint:`/info/<entry_type>` endpoint (see section `Entry Listing Info Endpoints`_)
+
+Definition-Provider-Specific Namespace Prefixes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This standard refers to definition-provider-specific prefixes and definition providers.
+
+Definition providers MUST provide a canonical property definition for all custom fields they define using the OPTIMADE `Property Definitions`_ format.
+Definition providers MUST also list these definitions in the relevant :endpoint:`/info/<entry_type>` endpoint of the index meta-database for that provider.
+They may also provide human-readable webpages for their definitions.
```
Comment (vaitkus) [inline optimade.rst:432]:
```suggestion
They MAY also provide human-readable webpages for their definitions.
```

File: optimade.rst [hunk 3]
```diff
@@ -367,26 +379,61 @@ A few suggestions and mandatory requirements of the OPTIMADE specification are s
...
+
+Definition-provider-specific fields MAY be fully described at the relevant :endpoint:`/info/<entry_type>` endpoint (see section `Entry Listing Info Endpoints`_) but can also rely on the canonical definitions provided by the definition provider, provided they return an ``$id`` for the field that resolves to the relevant OPTIMADE property definition.
```
Comment (vaitkus) [inline optimade.rst:434]:
```suggestion
Definition-provider-specific fields MAY be fully described at the relevant :endpoint:`/info/<entry_type>` endpoint (see section `Entry Listing Info Endpoints`_), but can also rely on the canonical definitions provided by the definition provider, provided they return an ``$id`` for the field that resolves to the relevant OPTIMADE property definition.
```

File: optimade.rst [hunk 4]
```diff
@@ -131,7 +131,19 @@ The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SH
     Every database provider is designated a unique prefix.
     The prefix is used to separate the namespaces used by provider-specific extensions.
     The list of presently defined prefixes is maintained externally from this specification.
-    For more information, see section `Database-Provider-Specific Namespace Prefixes`_.
+    For more information, see section `Namespace Prefixes`_.
+
+**Definition provider**
+    A service that provides one or more external or domain-specific property definitions that
```
Comment (merkys) [inline optimade.rst:137]: These two definitions do not follow our convention of having each sentence in a single line.

File: optimade.rst [hunk 5]
```diff
@@ -131,7 +131,19 @@ The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SH
...
+    can be used by OPTIMADE API implementations.
+
+**Definition provider prefix**
+    Every definition provider is designated a prefix that cannot clash with an
+    existing database provider prefix.
+    The prefix is used to separate the namespaces used by these collections of
+    definitions.
+    The list of presently defined prefixes is maintained externally from this specification.
+    For more information, see section `Namespace Prefixes`_.
```
Comment (rartino) [inline optimade.rst:146]:
```suggestion
**Definition provider**
    A service that provides one or more external or domain-specific property definitions that can be used by OPTIMADE API implementations.

**Definition provider prefix**
    Every definition provider is designated a prefix that cannot clash with an existing database provider prefix.
    The prefix is used to separate the namespaces used by these collections of definitions.
    The list of presently defined prefixes is maintained externally from this specification.
    For more information, see section `Namespace Prefixes`_.
```

**Comment by rartino** [comment]
@ml-evs I've gone through this, and aside the minor formatting issue @merkys remarked on (and I've provided a diff for), this looks very good to me and I don't quite see why we cannot already merge it.

The new [optimade-property-tools](https://github.com/Materials-Consortia/optimade-property-tools) repo with its example directory may make it a bit easier to run a namespace provider test.

File: optimade.rst [hunk 6]
```diff
@@ -131,7 +131,19 @@ The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SH
     Every database provider is designated a unique prefix.
     The prefix is used to separate the namespaces used by provider-specific extensions.
     The list of presently defined prefixes is maintained externally from this specification.
-    For more information, see section `Database-Provider-Specific Namespace Prefixes`_.
+    For more information, see section `Namespace Prefixes`_.
+
+**Definition provider**
+    A service that provides one or more external or domain-specific property definitions that
+    can be used by OPTIMADE API implementations.
```
Comment (ml-evs) [inline optimade.rst:138]:
```suggestion
    A service that provides one or more external or domain-specific property definitions that can be used by OPTIMADE API implementations.
```

**Comment by ml-evs** [comment]
Thanks @rartino, happy to merge this before rc.2 so we can test out the implications.

File: .words.lst [hunk 7]
```diff
@@ -214,4 +215,6 @@ xyz
 yacc
 yb
 zeo
-�ngstr�m
+ångström
+Ã¥ngstrÃ¶m
+
```
Comment (ml-evs) [inline .words.lst:220]:
```suggestion
```

**Review by rartino — APPROVED** [review]
This looks good to me. We are already trying this out in other repos.

File: optimade.rst [hunk 8]
```diff
@@ -367,26 +376,56 @@ A few suggestions and mandatory requirements of the OPTIMADE specification are s
 
 ..
 
-    **Note**: A list of database providers acknowledged by the **Open Databases Integration for Materials Design** consortium is maintained externally from this specification and can be retrieved as described in section `Database-Provider-Specific Namespace Prefixes`_.
-    This list is also machine-readable, optimizing the automatic discoverability.
+    **Note**: A list of database providers acknowledged by the **Open Databases Integration for Materials Design** consortium is maintained externally from this specification and can be retrieved as described in section `Namespace Prefixes`_.
```
Comment (merkys) [inline optimade.rst:379]: Shouldn't "database providers" here be updated to reflect that definition prefixes go there as well, or wasn't this the idea?

File: optimade.rst [hunk 9]
```diff
@@ -367,26 +376,56 @@ A few suggestions and mandatory requirements of the OPTIMADE specification are s
...
+
+This standard refers to definition-provider-specific prefixes and definition providers.
+
+Definition providers MUST provide a canonical property definition for all custom fields they define using the OPTIMADE `Property Definitions`_ format.
+Definition providers MUST also list these definitions in the relevant :endpoint:`/info/<entry_type>` endpoint of the index meta-database for that provider.
+They MAY also provide human-readable webpages for their definitions.
+
+Definition-provider-specific fields MAY be fully described at the relevant :endpoint:`/info/<entry_type>` endpoint (see section `Entry Listing Info Endpoints`_), but can also rely on the canonical definitions provided by the definition provider, provided they return an ``$id`` for the field that resolves to the relevant OPTIMADE property definition.
+
+
```
Comment (merkys) [inline optimade.rst:428]:
```suggestion
```

File: optimade.rst [hunk 10]
```diff
@@ -3390,13 +3429,14 @@ mtime
...
-Names of database-provider-specific entry types MUST start with database-provider-specific namespace prefix (see appendix `Database-Provider-Specific Namespace Prefixes`_).
-Database-provider-specific entry types MUST have all properties described above in section `Properties Used by Multiple Entry Types`_.
+Database and definition providers can define custom entry types.
+The names of such entry types MUST start with corresponding namespace prefix (see appendix `Namespace Prefixes`_).
+Custom entry types MUST have all properties described above in section `Properties Used by Multiple Entry Types`_.
 
-- **Requirements/Conventions for properties in database-provider-specific entry types**:
+- **Requirements/Conventions for properties in custom entry types**:
 
   - **Support**: Support for any properties in database-provider-specific entry types is fully OPTIONAL.
```
Comment (merkys) [inline optimade.rst:3441]: What about definition-provider-specific entry types wrt this? Do we want to say that they are OPTIONAL as well?

File: optimade.rst [hunk 11]
```diff
@@ -2515,10 +2554,10 @@ database-provider-specific properties
   - **Response**: API implementations are free to choose whether database-provider-specific properties are only included when requested using the query parameter :query-param:`response_fields`, or if they are included also when :query-param:`response_fields` is not present.
     Implementations are thus allowed to decide that some of these properties are part of what can be seen as the default value of :query-param:`response_fields` when that query parameter is omitted.
     Implementations SHOULD NOT include database-provider-specific properties when the query parameter :query-param:`response_fields` is present but does not list them.
-  - These MUST be prefixed by a database-provider-specific prefix (see appendix `Database-Provider-Specific Namespace Prefixes`_).
+  - These MUST be prefixed by a database-provider-specific prefix (see appendix `Namespace Prefixes`_).
```
Comment (merkys) [inline optimade.rst:2557]:
```suggestion
  - These MUST be prefixed by a database-provider-specific or definition-provider-specific prefix (see appendix `Namespace Prefixes`_).
```
There are two more occurrences in this very paragraph where "definition-provider-specific" seems to be missing, I just cannot suggest edits for them as they are too far from the changed lines.

**Review by merkys — COMMENTED** [review]
I made some minor suggestions to put definition-provider prefixes to the same level as database-provider prefixes. AFAIK they should be roughly equivalent from now.

File: optimade.rst (hunk 11 above)
Comment (ml-evs) [reply optimade.rst:2557]: In some of these cases, the database-provider specific "thing" is a query parameter or metadata field. I don't think there are any plans to create namespaces for these (or at least, we don't have the mechanism for this at the moment -- although e.g., `filter_smarts` might be one of these?) I think some of your comments refer to these cases but cannot easily verify right now.

File: optimade.rst (hunk 10 above)
Comment (ml-evs) [reply optimade.rst:3441]: wrt. to my other comment this change does make sense -- I think the proposed namespace provider format is flexible enough for custom entry types

File: optimade.rst (hunk 8 above)
Comment (ml-evs) [reply optimade.rst:379]:
```suggestion
    **Note**: A list of database and definition providers acknowledged by the **Open Databases Integration for Materials Design** consortium is maintained externally from this specification and can be retrieved as described in section `Namespace Prefixes`_.
```

File: optimade.rst (hunk 11 above)
Comment (ml-evs) [reply optimade.rst:2557]: Yeah I think this one should stay as it is; there is no way for a definition provider to overload query parameters yet.

File: optimade.rst (hunk 10 above)
Comment (ml-evs) [reply optimade.rst:3441]:
```suggestion
  - **Support**: Support for any properties in database-provider-specific or definition-provider-specific entry types is fully OPTIONAL.
```

File: optimade.rst (hunk 11 above)
Comment (merkys) [reply optimade.rst:2564]: Ah, right. I was overly eager to equate definition provider to database provider. Marking as resolved.

**Review by merkys — DISMISSED** [review]
Good to go, thanks!

**Review by rartino — DISMISSED** [review]
Re-confirming my positive approval.

**Comment by ml-evs** [comment]
Ok, now that the one linting issue is fixed (with the checker introduced since your last reviews...) I think this is the final time I need to bother you all with review requests (on this PR...)

**Review by rartino — APPROVED** [review]

**Review by merkys — APPROVED** [review]

**Pull request merged**